### ISSUE TITLE
BinaryAccuracy() sometimes gives incorrect answers due to non-deterministic sigmoiding 

### ISSUE THREAD

**Issue by idc9** [Lightning-AI/metrics#1604]
## 🐛 Bug

`torchmetrics.classification.BinaryAccuracy` will apply a sigmoid to some inputs but not others leading to incorrect behavior.

## Details
The current behavior of BinaryAccuracy() is to apply a sigmoid transformation if the inputs are outside of [0, 1] before binarizing   

> If preds is a floating point tensor with values outside [0,1] range we consider the input to be logits and will auto apply sigmoid per element. 

i.e. 
y_hat = 1(sigmoid(z) >= threshold) if z outside [0, 1] 
y_hat = 1(z >= threshold) if z inside [0, 1] 

I assume z inside [0, 1] is checked for then entire batch (i.e. if one element of the batch is outside [0, 1] then we apply the sigmoid to everyone).

This will cause silent errors. In particular, if the user inputs logits then they expect the logits to always be sigmoided. However, it is totally possible for all of the logits to lie in [0, 1] for some batches in which case the input will **not** be sigmoided which will cause incorrect thresholding.

### To Reproduce

Here is a simple example. Support our network outputs logits.

```python
from torchmetrics.classification import BinaryAccuracy
from scipy.special import expit # expit = sigmoid
import numpy as np
import torch
```

This example should lead to a correct prediction
``` python
probability_thresh = 0.5 
logits = np.array([0.49]) # network output
target = np.array([1])

# logits of 0.49 give a probability of 0.62 indicating class 1, the correct prediction
expit(logits)
array([0.62010643])
int(expit(logits) >= probability_thresh) == target
True
```

`BinaryAccuracy()` however thinks it's an incorrect prediction~
``` python
# torchmetrics, however, thinks we have the inccorect prediction because it does NOT sigmoid the logits
ba = BinaryAccuracy(threshold=probability_thresh) 
ba.forward(preds=torch.tensor(logits), target=torch.tensor(target))
tensor(0.)
```

### Suggested Fix

I suggest adding an argument indicating whether or not the input predictions are sigmoided so the inputs are either always sigmoided or never sigmoided

**Comment by idc9** [comment]
Looks like a similar thing happens in MultilabelAccuracy (https://torchmetrics.readthedocs.io/en/stable/classification/accuracy.html)

**Comment by JanFidor** [comment]
@Borda @SkafteNicki Hi, I came upon this issue when working on my engineering thesis and would very much like to help fix it if there's an opportunity (as a form of personal revenge on logits within [0, 1]). It seems that https://github.com/Lightning-AI/torchmetrics/pull/1676 was supposed to fix it, but appears to be abandoned. May I continue working on it?

**Comment by JanFidor** [comment]
Thanks for the quick signal @Borda ! I read the discussion on https://github.com/Lightning-AI/torchmetrics/pull/1676, question to all who contributed to the discussion @SkafteNicki @idc9 . I think that the solution propose above was a good direction, personally I don't like the idea of adding the 'auto' value for the sake of backwards compatibility. The bug is extremely hard to track down in the best of circumstances. After the fix with an option of 'auto' I'm afraid a lot of users could see their  code no longer working and using 'auto' without awareness of the inherent distortion to their metrics.  Right now it simply doesn't work well for small batches and it 'possibly' working for large ones doesn't feel like a good enough argument to keep it.

I'm open for discussion, but will defer to your judgment at the end of the day.
